core_api/src/lib.rs

```diff
@@ -79,14 +79,35 @@ impl Holochain {
         let name = dna.name.clone();
         let action = Nucleus(InitApplication(dna));
         instance.start_action_loop();
-        instance.dispatch_and_wait(action);
-        context.log(&format!("{} instantiated", name))?;
-        let app = Holochain {
-            instance,
-            context,
-            active: false,
-        };
-        Ok(app)
+
+        let (sender, receiver) = channel();
+
+        instance.dispatch_with_observer(action, move |state: &State| {
+            let nucleus_state = state.nucleus();
+            if nucleus_state.has_initialized() || nucleus_state.has_initialization_failed() {
+                sender
+                    .send(nucleus_state.status())
+                    .expect("test channel must be open");
+                true
+            } else {
+                false
+            }
+        });
+
+        // is unwrap ok here?
+        let status = receiver.recv().unwrap();
```

Comment:
Don't know for unwrap, but wouldn't you want to use `recv_timeout` instead in case nucleus' status never reach `initialized` or `initializedFail`